src/test/perl/path.t

```diff
@@ -277,63 +298,242 @@ $mc->{fail} = undef;
+    my ($mc, $target, $link_path, $expected_status) = @_;
+    my $noaction = $mc->_get_noaction();
+
+    my $ok_msg;
+    if ( $noaction ) {
+        $ok_msg = "$link_path hardlink not created (NoAction set)";
+    } else {
+        $ok_msg = "$link_path is " . ($expected_status == CHANGED ? "" : "already ") . "a hardlink";
+    }
+    my $target_msg = "$link_path hardlink has the expected " . ($expected_status == CHANGED ? "changed " : "") . "target ($target)";
+
+    # Test if symlink was created or not according to NoAction flag
+    my $ok_condition;
+    if ( $noaction ) {
+        $ok_condition = ! $mc->any_exists($link_path);
+    } else {
+        $ok_condition = $mc->has_hardlinks($link_path);
+    }
+    ok($ok_condition, $ok_msg);
+
+    ok($mc->is_hardlink($link_path, $target), $target_msg) unless $noaction;
+};
+
+init_exception("hardlink tests");
+
+rmtree ($basetest) if -d $basetest;
+my $cwd = cwd();
+my $hardlink1 = "$basetest/$target_directory/hardlink1";
+my $hardlink2 = "$basetest/$target_directory/hardlink2";
+# hardlink target must be an absolute path or the link path directory is prepended
+$target_file1 = "$cwd/$basetest/tgtfile1";
+my $relative_target_file3 = "$basetest/$target_directory/tgtfile3";
+my $target_file3 = "$cwd/$relative_target_file3";
+makefile($target_file1);
+makefile($target_file3);
+
+for $CAF::Object::NoAction (1,0) {
+    is($mc->hardlink($target_file1, $hardlink1), CHANGED, "hardlink created");
+    check_hardlink($mc, $target_file1, $hardlink1, CHANGED);
+    if ( !$CAF::Object::NoAction ) {
+        is($mc->hardlink($target_file1, $hardlink1), SUCCESS, "hardlink already exists");
+        check_hardlink($mc, $target_file1, $hardlink1, SUCCESS);
+    }
+    is($mc->hardlink($target_file3, $hardlink1), CHANGED, "hardlink updated");
+    check_hardlink($mc, $target_file3, $hardlink1, CHANGED);
+}
+
+# The following tests are not affected by NoAction flag
+$link_status = $mc->hardlink("missing_file", $hardlink1);
+ok(! $link_status, "hardlink not created (target has to exist)");
+is($mc->{fail},
+   "*** invalid target (missing_file): stat($basetest/$target_directory/missing_file): No such file or directory",
+   "fail attribute set after hardlink error");
+$link_status = $mc->hardlink($target_file1, "$basetest/$target_directory");
+ok(! $link_status, "hardlink not created (do not replace an existing directory)");
+is($mc->{fail},
+   "*** cannot hard link $basetest/$target_directory: it is a directory",
+   "fail attribute set after hardlink error (existing directory not replaced)");
+$link_status = $mc->hardlink($relative_target_file3, $hardlink2);
+ok(! $link_status, "hardlink not created (relative target, link path directory prepended)");
+is($mc->{fail},
+   "*** invalid target ($relative_target_file3): stat(". dirname($hardlink2) . "/$relative_target_file3): No such file or directory",
+   "fail attribute set after hardlink error (relative target, link path directory prepended)");
+
```

Comment:
As stated in the comment, `check_hardlink()` is the main unit test for these methods. I don't see the point of adding something specific that will do exactly the same thing (we need to create something and check that it works)... I'll also add these for cases where the argument is not a hardlink in the previous section of tests.